sql/2024/fonts/design/fonts_metric.sql

```diff
@@ -0,0 +1,133 @@
+-- Section: Performance
+-- Question: What is the distribution of metrics?
+-- Normalization: Fonts (parsed only)
+
+-- INCLUDE ../common.sql
+
+WITH
+fonts AS (
+  SELECT
+    client,
+    url,
+    SAFE_CAST(
+      JSON_EXTRACT_SCALAR(
+        ANY_VALUE(payload),
+        '$._font_details.head.unitsPerEm'
+      ) AS INTEGER
+    ) AS granularity,
+    [
+      STRUCT(
+        'granularity' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.head.unitsPerEm'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'clipping_ascender' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.usWinAscent'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'ascender' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.sTypoAscender'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'cap_height' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.sCapHeight'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'x_height' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.sxHeight'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'descender' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.sTypoDescender'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'clipping_descender' AS name,
+        -SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.usWinDescent'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'line_gap' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.sTypoLineGap'
+          ) AS INTEGER
+        ) AS value
+      ),
+      STRUCT(
+        'use_typographic_metrics' AS name,
+        SAFE_CAST(
+          JSON_EXTRACT_SCALAR(
+            ANY_VALUE(payload),
+            '$._font_details.OS2.fsSelection'
+          ) AS INTEGER
+        ) & 128 AS value
+      )
+    ] AS metrics
+  FROM
+    `httparchive.all.requests`
+  WHERE
+    date = '2024-07-01' AND
+    type = 'font' AND
+    is_root_page AND
+    IS_PARSED(payload)
+  GROUP BY
+    client,
+    url
+)
+
+SELECT
+  client,
+  name AS metric,
+  percentile,
+  COUNT(0) AS count,
+  APPROX_QUANTILES(
+    IF(name = 'granularity', value, SAFE_DIVIDE(value, granularity)),
+    1000
+  )[OFFSET(percentile * 10)] AS value
+FROM
+  fonts,
+  UNNEST(metrics) AS metric,
+  UNNEST([10, 25, 50, 75, 90, 99]) AS percentile
```

Comment:
As discussed in chat would 100 be better than 99? It’s kinda meaningless at this extremes percentile anyway but highlighting the “worst use” can sometimes lead to interesting insights whereas I don’t think there’s anything of interest in p99 (since it’s highly volatile but also not that worse case).

Reply:
I think everything above the 90th will be ignored in the body of the chapter. I lowered it just to have something more sensible there; there will always be a handful of corky fonts with unrealistic parameters set.